pkg/adaptor/proxy/service.go

```diff
@@ -103,38 +108,81 @@ func (s *proxyService) redirect(ctx context.Context, fn func(c *client)) {
 	fn(s.agentClient)
 }
 
-// TODO: parameterize the pause container image name
-const pauseContainerImage = "k8s.gcr.io/pause:3.7"
+func (s *proxyService) initCriClient(ctx context.Context) error {
+	if s.criClient != nil {
+		return nil
+	}
 
-func getImageFromDigest(digest string) (string, error) {
-	cmd := exec.Command("bash", "-c", "/usr/local/bin/crictl inspecti "+digest+" | /usr/bin/jq '.status.repoTags'")
-	out, err := cmd.CombinedOutput()
-	if err != nil {
-		return "", err
+	if s.criSocketPath != "" {
+		maxRetries := s.maxRetries
+		count := 1
+		for {
+			var err error
+			func() {
+				ctx, cancel := context.WithTimeout(ctx, s.retryInterval)
+				defer cancel()
+
+				s.criClient, err = containerd.New(s.criSocketPath)
+				if err == nil || count == maxRetries {
+					return
+				}
+				<-ctx.Done()
+			}()
+
+			if err == nil {
+				break
+			}
+			if count == maxRetries {
+				s.criClient = nil
+				err := fmt.Errorf("reaches max retry count. gave up establishing cri uds connection to %s: %w", s.criSocketPath, err)
+				logger.Print(err)
+				return err
+			}
+			logger.Printf("failed to establish cri uds connection to %s: %v. (retrying... %d/%d)", s.criSocketPath, err, count, maxRetries)
+
+			count++
+		}
+
+		s.criImageClient = runtime.NewImageServiceClient(s.criClient.Conn())
+		logger.Printf("established cri uds connection to %s", s.criSocketPath)
+		return nil
+	}
+
+	return fmt.Errorf("cri runtime endpoint is not specified, it is used to get the image name from image digest.")
+}
+
+// TODO: REMOVE ME AFTER cc supported pass image name rather than digest for existing image
+func (s *proxyService) getImageFromDigest(ctx context.Context, digest string) (string, error) {
+	if s.criClient == nil {
+		err := s.initCriClient(ctx)
+		if err != nil {
+			return "", err
+		}
```

Comment:
Race here. How about using `sync.Once` instead?

Reply:
Good point, there is a race here indeed.

Reply:
I talked with Lei this morning, he tends to use this cri grpc client also for CSI works. so, looks we can move it to initialization algorithm, did the change accordingly.